### PR TITLE
TOOLS-1855 allow multiple services per resource

### PR DESCRIPTION
**Pull request by grosser**
@sandlerr @jonmoter @irwaters 

```
kube-service-watcher            None             <nodes>       80:30653/TCP                                                  1m
kube-service-watcher2           None             <nodes>       81:30291/TCP                                                  1m
```

### CODE REVIEW

File: plugins/kubernetes/app/models/kubernetes/role_verifier.rb [hunk 1]
```diff
@@ -62,8 +62,15 @@ def verify_namespace
     end
 
     def verify_kinds
-      kinds = map_attributes([:kind]).sort_by(&:to_s)
-      return if SUPPORTED_KINDS.include?(kinds - ['ConfigMap'])
+      kinds = map_attributes([:kind])
+      kinds -= ['ConfigMap']
+      if kinds.count('Service') > 1
+        kinds -= ['Service']
+        kinds += ['Service']
```
Comment (henders) [inline plugins/kubernetes/app/models/kubernetes/role_verifier.rb]: This is confusing :)

File: plugins/kubernetes/app/models/kubernetes/release_doc.rb [hunk 2]
```diff
@@ -92,6 +85,23 @@ def store_resource_template
...
+          Samson::Hooks::UserError,
+          "Service name for role #{kubernetes_role.name} was generated and needs to be changed before deploying."
+        )
+      end
+
+      return config_name if config_name.start_with?(name)
+
+      @service_names_generated ||= 0
+      @service_names_generated += 1
+      name += "-#{@service_names_generated}" if @service_names_generated >= 2
```
Comment (henders) [inline plugins/kubernetes/app/models/kubernetes/release_doc.rb]: nit: `if @service_names_generated > 1` seems a bit more intuitive. But either way.

**Comment by grosser** [comment]
hehe ... if there is more then 1 service consider it 1 service ... for
comparison with supported configurations

File: plugins/kubernetes/app/models/kubernetes/role.rb [hunk 3]
```diff
@@ -58,7 +58,7 @@ def self.seed!(project, git_ref)
         name = resource.fetch(:metadata).fetch(:labels).fetch(:role)
 
         # service
-        if service = config_file.service
+        if service = config_file.services.first
```
Comment (henders) [inline plugins/kubernetes/app/models/kubernetes/role.rb]: I'm assuming it doesn't matter that we don't show all the services in the role UI.

**Comment by grosser** [comment]
that's a bit weird ... since we only have 1 field for service-name I'm
using it and then adding a suffix ... kinda shady ...

File: plugins/kubernetes/app/models/kubernetes/role_verifier.rb (hunk 1 above)
Comment (henders) [reply plugins/kubernetes/app/models/kubernetes/role_verifier.rb]: Oh, you're just trying to remove duplicates?

**Comment by henders** [comment]
Cool, I think it's grand anyway if the UI lags a bit behind for this, I don't know who uses the UI to look at the service name anyway.

👍

**Comment by grosser** [comment]
yeah, but only for Service ... so it's a bit ugly :/

**Comment by grosser** [comment]
added more docs for the weird parts

**Comment by klaus993** [comment]
Hi @grosser 

I have an issue with some of this code (though I'm using `v1938`, this code is the same but in template_filler). I'm trying to deploy only one role, its yaml defines a deployment with three containers and only one service:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: app-name
  labels:
    project: app-name
    role: web
spec:
  type: NodePort
  ports:
  - name: http
    port: 80
    protocol: TCP
  selector:
    project: app-name
    role: web
```

When deploying, after the Docker build is successful, this is the error in the Samson console:

```
[04:29:43] JobExecution failed: Service name for role web was generated and needs to be changed before deploying.
[04:29:43] plugins/kubernetes/app/models/kubernetes/template_filler.rb:103:in `generate_service_name'
[04:29:43] plugins/kubernetes/app/models/kubernetes/template_filler.rb:91:in `set_service_name'
[04:29:43] plugins/kubernetes/app/models/kubernetes/template_filler.rb:25:in `to_hash'
[04:29:43] plugins/kubernetes/app/models/kubernetes/release_doc.rb:77:in `block in store_resource_template'
[04:29:43] plugins/kubernetes/app/models/kubernetes/release_doc.rb:75:in `map'
[04:29:43] plugins/kubernetes/app/models/kubernetes/release_doc.rb:75:in `store_resource_template'
[04:29:43] plugins/kubernetes/app/models/kubernetes/release.rb:80:in `block (2 levels) in create_release_docs'
[04:29:43] plugins/kubernetes/app/models/kubernetes/release.rb:79:in `each'
[04:29:43] plugins/kubernetes/app/models/kubernetes/release.rb:79:in `block in create_release_docs'
[04:29:43] plugins/kubernetes/app/models/kubernetes/release.rb:78:in `each'
```

I don't understand, what does `Service name for role web was generated and needs to be changed before deploying` mean? This is deployed with a previous version of Samson and there is no problem.

Thank you in advance :)

**Comment by grosser** [comment]
check the kubernetes roles UI ... the name there will have `generated-` ... replace that with something real like `my-app-server`

**Comment by klaus993** [comment]
Thank you! I had not seen that, sorry. So it's only a matter of names :)

**Pull request merged**